Library/Homebrew/rubocops/lines.rb

```diff
@@ -517,14 +524,64 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
 
         # matches ({bash,zsh,fish}_completion/"_?foo{.fish}?").write output
         def_node_search :shell_completion_node, <<~EOS
-        $(send
-          (begin
-            (send
-              (send nil? {:bash_completion :zsh_completion :fish_completion}) :/
-              (str _))) :write _)
+          $(send
+            (begin
+              (send
+                (send nil? {:bash_completion :zsh_completion :fish_completion}) :/
+                (str _))) :write _)
         EOS
       end
 
+      # This cop makes sure that the `generate_completions_from_executable` DSL is used with only
+      # a single, combined call for all shells.
+      #
+      # @api private
+      class SingleGenerateCompletionsDSLCall < FormulaCop
+        extend AutoCorrector
+
+        def audit_formula(_node, _class_node, _parent_class_node, body_node)
+          install = find_method_def(body_node, :install)
+
+          methods = find_every_method_call_by_name(install, :generate_completions_from_executable)
+          return if methods.length <= 1
+
+          offenses = []
+          shells = []
+          methods.each do |method|
+            next unless method.source.include?("shells:")
+
+            shells << method.source.match(/shells: \[(:bash|:zsh|:fish)\]/).captures.first
+            offenses << method
+          end
+
+          return if offenses.blank?
+
+          offenses[0...-1].each do |node|
+            offending_node(node)
+            problem "Use a single `generate_completions_from_executable` call
+                         combining all specified shells" do |corrector|
+              corrector.replace(@offensive_node.source_range, "")
+            end
+          end
+
+          offending_node(offenses.last)
+          replacement = if (shells - %w[:bash :zsh :fish]).empty?
+            @offensive_node.source.sub(/shells: \[(:bash|:zsh|:fish)\]/, "")
+                           .gsub(",,", ",")
+                           .sub(", )", ")")
+                           .sub("(, ", "(")
+                           .sub("()", "")
```

Comment:
What do all of these do?

Reply:
Clean up the replacement string after updating the `shells` argument (because I did not figure these cases out in the regex, and I don't think that's even possible):
1. might not be needed, on second thought
2. `shells` was the last argument, the substitution would leave a trailing comma behind
3. `shells` was the first argument, the substitution would leave a leading comma in the front
4. `shells` was the only argument, the substitution would leave empty parentheses behind

Reply:
What about something like this:

```ruby
            @offensive_node.source.sub(/(,\s*)?shells: \[(:bash|:zsh|:fish)\]/, "").sub(/,\s*\)/, "")
```

That will include the comma before, and then we only have to deal with the case of the comma after.

I'm pretty sure there's already a rubocop for empty parentheses that will take care of that for you (would be good to double check, though)

You could also use a single regex with `(,\s*)?` on both ends and then do some more manual parsing, maybe by passing a block to `sub`. Not 100% sure. I think a comment here would be useful if you want to stick with this, though

Reply:
> ```ruby
>             @offensive_node.source.sub(/(,\s*)?shells: \[(:bash|:zsh|:fish)\]/, "").sub(/,\s*\)/, "")
> ```

This did not quite work. But you were right on the cop cleaning up parentheses!
I've removed the unnecessary cases and added comments for the remaining two.